x-pack/plugins/fleet/server/services/policy_preconfig.ts

```diff
@@ -0,0 +1,198 @@
+    // package-a:1.0.0, package-a:2.0.0; package-b:1.0.0, package-b:2.0.0
+    const duplicateList = duplicatePackages
+      .map(([, versions]) => versions.map((v) => `${v.name}:${v.version}`).join(', '))
+      .join('; ');
+
+    throw new Error(
+      i18n.translate('xpack.fleet.setup.duplicatePackageError', {
+        defaultMessage: 'Duplicate packages specified in configuration: {duplicateList}',
+        values: {
+          duplicateList,
+        },
+      })
+    );
+  }
+
+  // Preinstall packages specified in Kibana config
+  const preconfiguredPackages = await Promise.all(
+    packages.map(({ name, version }) =>
+      ensureInstalledPreconfiguredPackage(soClient, esClient, name, version)
+    )
+  );
+
+  // Create policies specified in Kibana config
+  const preconfiguredPolicies = await Promise.all(
+    policies.map(async ({ package_policies: packagePolicies, id, ...newAgentPolicy }) => {
+      const { created, policy } = await agentPolicyService.ensurePreconfiguredAgentPolicy(
+        soClient,
+        esClient,
+        { ...newAgentPolicy, preconfiguration_id: String(id) }
+      );
+
+      if (!created) return { created, policy };
+
+      const installedPackagePolicies = await Promise.all(
+        packagePolicies.map(async ({ package: pkg, name, ...newPackagePolicy }) => {
+          const installedPackage = await isPackageInstalled({
+            savedObjectsClient: soClient,
+            pkgName: pkg.name,
+          });
+          if (!installedPackage) {
+            throw new Error(
+              i18n.translate('xpack.fleet.preconfiguredPackageMissingError', {
+                defaultMessage:
+                  '{agentPolicyName} could not be added. {pkgName} is not installed, add {pkgName} to `{packagesConfigValue}` or remove it from {packagePolicyName}.',
+                values: {
+                  agentPolicyName: newAgentPolicy.name,
+                  packagePolicyName: name,
+                  pkgName: pkg.name,
+                  packagesConfigValue: 'xpack.fleet.packages',
+                },
+              })
+            );
+          }
+          return { name, installedPackage, ...newPackagePolicy };
+        })
+      );
+
+      return { created, policy, installedPackagePolicies };
+    })
+  );
+
+  for (const preconfiguredPolicy of preconfiguredPolicies) {
+    const { created, policy, installedPackagePolicies } = preconfiguredPolicy;
+    if (created) {
+      await addPreconfiguredPolicyPackages(
+        soClient,
+        esClient,
+        policy,
+        installedPackagePolicies!,
+        defaultOutput
+      );
+    }
+  }
+
+  return {
+    policies: preconfiguredPolicies.map((p) => ({
+      id: p.policy.id,
+      updated_at: p.policy.updated_at,
+    })),
+    packages: preconfiguredPackages.map((pkg) => `${pkg.name}:${pkg.version}`),
+  };
+}
+
+export async function addPackageToAgentPolicy(
```

Comment:
rather than moving this here from `services/setup.ts`, let's just duplicate most of the code, with my suggestions in `package_to_package_policy.ts` (we can just add the duplicated logic into `addPreconfiguredPolicyPackages` for example)